src/main/java/co/cask/http/internal/RequestRouter.java

```diff
@@ -0,0 +1,163 @@
+ */
+
+public class RequestRouter extends ChannelInboundHandlerAdapter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HttpDispatcher.class);
+
+  private final int chunkMemoryLimit;
+  private final HttpResourceHandler httpMethodHandler;
+  private final AtomicBoolean exceptionRaised;
+  private final boolean sslEnabled;
+
+  private HttpMethodInfo methodInfo;
+
+  public RequestRouter(HttpResourceHandler methodHandler, int chunkMemoryLimit, boolean sslEnabled) {
+    this.httpMethodHandler = methodHandler;
+    this.chunkMemoryLimit = chunkMemoryLimit;
+    this.exceptionRaised = new AtomicBoolean(false);
+    this.sslEnabled = sslEnabled;
+  }
+
+  /**
+   * If the HttpRequest is valid and handled it will be sent upstream, if it cannot be invoked
+   * the response will be written back immediately.
+   */
+  @Override
+  public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+    if (exceptionRaised.get()) {
+      return;
+    }
+    if (!(msg instanceof HttpRequest)) {
+      // If there is no methodInfo, it means the request was already rejected.
+      // What we received here is just residue of the request, which can be ignored.
+      if (methodInfo != null) {
+        super.channelRead(ctx, msg);
+      }
+      return;
+    }
+    HttpRequest request = (HttpRequest) msg;
+    BasicHttpResponder responder = new BasicHttpResponder(ctx.channel(), sslEnabled);
+    methodInfo = prepareHandleMethod(request, responder, ctx);
+
+    if (methodInfo != null) {
+      ctx.fireChannelRead(msg);
+    } else {
+      if (!responder.isResponded()) {
+        // If not yet responded, just respond with a not found and close the connection
+        HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_FOUND);
+        ctx.channel().writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
+      } else {
+        // If already responded, just close the connection
+        ctx.channel().close();
+      }
+    }
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+    String exceptionMessage = "Exception caught in channel processing.";
+    if (exceptionRaised.compareAndSet(false, true)) {
+      if (methodInfo != null) {
+        LOG.error(exceptionMessage, cause);
+        methodInfo.sendError(HttpResponseStatus.INTERNAL_SERVER_ERROR, cause);
+        methodInfo = null;
+      } else {
+        if (cause instanceof HandlerException) {
+          HttpResponse response = ((HandlerException) cause).createFailureResponse();
+          // trace logs for user errors, error logs for internal server errors
+          if (isUserError(response)) {
+            LOG.trace(exceptionMessage, cause);
+          } else {
+            LOG.error(exceptionMessage, cause);
+          }
+          ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
```

Comment:
I think this comment applies here: https://github.com/caskdata/netty-http/pull/46#discussion_r143300495